### PR TITLE
Fixes #16501: Optimize Rows Sample Type Snowflake

### LINKED ISSUE #16501: When using "Rows" for "Profile Sample Type", Snowflake Profiler is inefficient
**Affected module**
Ingestion Framework

**Describe the bug**
When using "Rows" for "Profile Sample Type" in Profiler Ingestion, the ORM does not leverage Snowflake's SAMPLE feature, but instead creates a CTE using `WITH "TABLE_NAME_rnd" AS ( SELECT … MOD(0, xxxxx) AS random …. ORDER BY random LIMIT XXXX) …`.

This is very inefficient on very large tables, as it must perform the `MOD(0, xxxxx) as random` bit against every single row in the table before it can perform the pseudo sort (on a column named "random" which all consist of same value). In one instance, we have 80B+ records in a table, and it took over 8 hours to compute for every record on a medium sized Snowflake warehouse.

Switching it over to "Percentage"  for "Profile Sample Type" results in the CTE using `SAMPLE BERNOULLI`, which is an improvement, but still not ideal, as it must perform the calculation against large chunk of the data. Going back to the 80B+ records table example, when using 5%, it must compute `MOD(0, 100)` against 4B records. However, it was able to complete the request in 26 minutes on the same medium sized Snowflake warehouse.

Because we have table of varying size (some as small as 1000 records) in this schema, configuring to 1% sample is not an ideal solution. 

If we can set 1M records as sample threshold, and a path forward for this would be to leverage the  `TABLESAMPLE (XXXXX ROWS)` query in Snowflake in the `_rnd` CTE block. By leveraging the `TABLESAMPLE (XXXXX ROWS)` in the `_rnd` CTE, it would only need to perform the `MOD(0, XXXXX)` on the configured records (i.e.: 1M vs 4B). This query took 22 minutes to run on the same medium sized Snowflake warehouse. 

I do not fully grasp the requirement for `MOD(0, XXXXX)` bit, but even with the `MOD(0, XXXXX)` bit in place, I can't help but to wonder if this bit could be removed, but that might be another analysis at another time.

All test were done with warehouse from cold start with at least 15 minutes between suspend and wake up.

**To Reproduce**

1. Create a large table with a lot of data.
2. Configure "Profile Sample Type" to "Percentage", observe relatively long but acceptable run time;
3. Configure "Profile Sample Type" to "Rows", observe a much longer run time;
4. Test ORM generated query with minimal modification, observe relatively long but acceptable run time

**Expected behavior**

I wish to see "Rows" based "Profile Sample Type" to run without needing to load the entire table before discarding.

**Version:**
 - OS: Docker OMD 1.4.0

**Additional context**
Reference Slack Chat: https://openmetadata.slack.com/archives/C02B6955S4S/p1717185445543959

### PR DESCRIPTION
**Pull request by ayush-shah**
### Describe your changes:

Fixes #16501

I worked on optimizing row sample type in snowflake

#
### Type of change:

- [ ] Bug fix
- [x] Improvement
- [ ] New feature
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation

#
### Checklist:

- [x] I have read the [**CONTRIBUTING**](https://docs.open-metadata.org/developers/contribute) document.
- [x] My PR title is `Fixes <issue-number>: <short explanation>`
- [x] I have commented on my code, particularly in hard-to-understand areas. 
- [x] For JSON Schema changes: I updated the migration scripts or explained why it is not needed.

### CODE REVIEW

File: ingestion/src/metadata/profiler/processor/sampler/sqlalchemy/snowflake/sampler.py
Comment (TeddyCr) [inline ingestion/src/metadata/profiler/processor/sampler/sqlalchemy/snowflake/sampler.py:1]: @ayush-shah can you move the one for the percentage out of the core sampler into here?

Comment (ayush-shah) [reply ingestion/src/metadata/profiler/processor/sampler/sqlalchemy/snowflake/sampler.py:1]: Modified the same, Thanks 🙏

File: ingestion/src/metadata/profiler/processor/sampler/sqlalchemy/snowflake/sampler.py [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    def get_sample_query(self, *, column=None) -> CTE:
+        """get query for sample data"""
+        # TABLESAMPLE SYSTEM is not supported for views
+        self.table = cast(Table, self.table)
+
+        if self.profile_sample_type == ProfileSampleType.PERCENTAGE:
+            rnd = (
+                self._base_sample_query(
+                    column,
+                    (ModuloFn(RandomNumFn(), 100)).label(RANDOM_LABEL),
```
Comment (TeddyCr) [inline ingestion/src/metadata/profiler/processor/sampler/sqlalchemy/snowflake/sampler.py:50]: I don't think we need that anymore here as the sampling will be done from `SAMPLE BER....`

File: ingestion/src/metadata/profiler/processor/sampler/sqlalchemy/snowflake/sampler.py [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+                )
+                .suffix_with(
+                    f"SAMPLE BERNOULLI ({self.profile_sample or 100})",
+                )
+                .cte(f"{self.table.__tablename__}_rnd")
+            )
+            session_query = self.client.query(rnd)
+            return session_query.where(rnd.c.random <= self.profile_sample).cte(
+                f"{self.table.__tablename__}_sample"
+            )
```
Comment (TeddyCr) [inline ingestion/src/metadata/profiler/processor/sampler/sqlalchemy/snowflake/sampler.py:60]: Same thing on the random logic

**Review by TeddyCr — APPROVED** [review]

**Pull request merged**